src/integrations/integr_chrome.rs

```diff
@@ -150,84 +159,93 @@ async fn start_chrome_session(
             .ok_or(format!("Error getting chrome session for chat: {}", session_hashmap_key))?
             .clone()
     };
-
+    
     let mut command_session_locked = command_session.lock().await;
     let chrome_session = command_session_locked.as_any_mut().downcast_mut::<ChromeSession>().ok_or("Failed to downcast to ChromeSession")?;
 
+    let mut tool_log = vec![];
+    let tab = match chrome_session.tabs.get(tab_name) {
+        Some(tab) => {
+            tool_log.push(format!("Using opened tab {}\n", tab_name.clone()));
+            tab.clone()
+        },
+        None => {
+            let tab = chrome_session.browser.new_tab().map_err(|e| e.to_string())?;
+            chrome_session.tabs.insert(tab_name.clone(), tab.clone());
+            tool_log.push(format!("Opened new tab {}\n", tab_name.clone()));
+            tab
+        }
+    };
+
     let mut messages = vec![];
     if command_args[0] == "navigate_to" {
         if command_args.len() < 2 {
-            messages.push(tool_message(format!("Missing argument `url`: {:?}", command_args), tool_call_id));
+            tool_log.push(format!("Missing required argument `uri`: {:?}. Call this command another time using format `navigate_to <uri>`.", command_args));
         } else {
-            chrome_session.tab.navigate_to(command_args[1].as_str()).map_err(|e| e.to_string())?;
-            chrome_session.tab.wait_until_navigated().map_err(|e| e.to_string())?;
-            messages.push(tool_message(format!("Chrome tab navigated to {}", command_args[1]), tool_call_id));
+            let content = navigate_to(&tab, &command_args[1]).await.map_err(
+                |e| format!("Cannot navigate_to `{}`: {}. Probably url doesn't exist. If you're trying to open local file add file:// prefix.", command_args[1], e)
+            )?;
+            tool_log.push(content);
         }
     } else if command_args[0] == "screenshot" {
-        messages.push(tool_message("Made a screenshot".to_string(), tool_call_id));
-        let screenshot_message = screenshot_jpeg_base64(&chrome_session.tab, false).await?;
-        messages.push(ContextEnum::ChatMessage(screenshot_message));
+        force_screenshot = true;
     } else if command_args[0] == "html" {
-        let content: String;
         let client = Client::builder()
             .build()
             .map_err(|e| e.to_string())?;
-        let url = chrome_session.tab.get_url();
+        let url = tab.get_url();
         let response = client.get(url.clone()).send().await.map_err(|e| e.to_string())?;
         if !response.status().is_success() {
-            content = format!("Unable to fetch url: {}; status: {}", url, response.status());
+            tool_log.push(format!("Unable to fetch url: {}; status: {}", url, response.status()));
         } else {
-            content = response.text().await.map_err(|e| e.to_string())?;
+            tool_log.push(response.text().await.map_err(|e| e.to_string())?);
```

Comment:
I'd recommend putting more simple facts about the web page into tool logs. 

Sent https://finance.yahoo.com/ to tab 1
Completed operation in 0.4s
Current tabs:

[
{"tab_index": 0, "url": "https://www.google.com/", "mobile": true},
{"tab_index": 1, "url": "https://finance.yahoo.com/"},
{"tab_index": 2, "url": "http://EpicFails.com"},
]